README.md
# About ZSS

ZSS is a small, simple, pure Lua library that parses simple CSS (see below for limitations) rules and then computes the declarations to apply for a given 'element'.

## Features:

* There is no assumption about the set of element types (names), ids, classes, or attributes allowed.
* There is no enforcement about the set of legal declaration names or values.
* It supports handlers to process "at-rules" as they are seen.
* It uses arbitrary Lua expressions for property values, with a sandbox table for resolving values and functions.
  * _For example, you can make_ `{ fill:none }` _in CSS turn into_ `{fill=false}` _in Lua._
  * _For example, you can make_ `{ fill:hsv(137, 0.5, 2.7) }` _in CSS turn into_ `fill = {r=0, g=2.7, b=0.765}` _in Lua._
* Value expressions can use placeholder values that are deferred and resolved against data supplied with a query. For example:
    ```lua
    local C = require'color'

    local style=zss:new():constants{
      white = C.white,
      hsv   = function(h,s,v) return C:new{h=h,s=s,v=v} end
    }:add[[
      circle          { fill:white; r:3      }
      circle[@r]      { radius:@r            }
      circle[@hue]    { fill:hsv(@hue, 1, 1) }
    ]]

    local props = style:match{type='circle'}
    inspect(props)
    --> { fill=<color 'white'>, radius=3 }

    local rnd=math.random
    for r=1,4 do
      local props = style:match{type='circle', r=r, hue=rnd(360) }
      inspect(props)
    end
    --> { fill=<color r=1.0 g=0.0 b=1.0 a=1.0>, radius=1 }
    --> { fill=<color r=0.0 g=1.0 b=0.4 a=1.0>, radius=2 }
    --> { fill=<color r=0.7 g=0.0 b=1.0 a=1.0>, radius=3 }
    --> { fill=<color r=0.8 g=0.0 b=1.0 a=1.0>, radius=4 }
    ```


## Simplifications/Limitations:

* ZSS assumes a **flat, unordered data model**. This means that there are no hierarchical selectors (no descendants, no children), no sibling selectors, and no pseudo-elements related to position.
* Using Lua to parse property expressions means there is no support for custom units on numbers using CSS syntax, or hexadecimal colors. You would need to wrap these in function calls like `len(5,cm)` or `color('#ff0033')`.
* Due to a simple parser:
  * you must not have a `{` character inside a selector. (Then again, when would you?)
  * you must not have a `;` character inside your declarations.
  * you must not have a `@` character inside your declarations.
  * you must not have an _unpaired_ `}` character inside your declarations.

The following CSS is currently unparsable in many ways:

```css
boo[a=='{'] { str:'nope'         } /* The { causes the selector to stop being parsed too soon    */
boo2        { no1:';'; ok:1      } /* The ; causes the value to stop being parsed too soon       */
boo3        { email:'me@here';   } /* The value get destroyed and becomes 'me_data_.here'        */
boo4        { no2:'}'            } /* The } causes the declaration to stop being parsed too soon */
yay         { tbl:{ a={b='ok'} } } /* Having { and } paired is OK, however                       */
```

If you need to have one of these characters inside a string, you can escape them using their hexadecimal escape equivalent:

* `;` — \x3B
* `@` — \x40
* `{` — \x7B
* `}` — \x7D

```css
person   { text:'me\x40company.com'       } /* me@company.com    */
message  { text:'tl\x3Bdr: this works'    } /* tl;dr: this works */
mustache { normal:':\x7B'; wicked:':\x7D' } /*    :{     :}      */
```


## Example Usage

```lua
ZSS   = require'zss'
color = require'color'

local style = ZSS:new{
  constants = {
    none        = false,
    transparent = false,
    visible     = true,
    rgb=function(r,g,b) return color{ r=r, g=g, b=b } end,
    hsv=function(h,s,v) return color{ h=h, s=s, v=v } end
  }
}

-- load additional constants into the data model for resolution
-- before parsing any CSS that might use them
style:constants(color.predefined)

style:add [[
  @font-face { font-family:'main'; src:'Arial.ttf' }
  @font-face { font-family:'bold'; src:'Arial-Bold.ttf' }
  *    { fill:none; stroke:white; size:20 }
  text { font:'main'; fill:white; stroke:none }
]]

style:load('my.css')
--> my.css has the contents
-->
--> @font-face      { font-family:'main'; src:'DINPro-Light.ttf' }
--> @font-face      { font-family:'bold'; src:'DINPro-Medium.ttf' }
--> #title          { font:'bold' }
--> .important      { fill:hsv(60,1,1) }
--> .debug, .boring { fill:white; opacity:0.2 }
--> text.debug,
--> text.boring     { opacity:0.5 }
--> text[@speed>20] { fill:red; opacity:1 }

style:match{ type='line' }
--> {fill=false, size=20, stroke=<color 'white'>}

style:match{ type='text' }
--> {fill=<color 'white'>, font="main", size=20, stroke=false}

style:match{ type='text', tags={important=1} }
--> {fill={1, 1, 0}, font="main", size=20, stroke=false}

style:match{ id='title' }
--> {fill={1, 1, 0}, font="bold", size=20, stroke=<color 'white'>}

style:match{ type='text', tags={debug=1}, speed=37 }
--> {fill=<color 'red'>, font="main", opacity=1, size=20, stroke=false}
```


# Per-sheet Variables

ZSS adds a custom `@vars` directive handler that allows each sheet loaded into a style to have its own set of variables expressed as declarations, that can be used in other rules. For example:

```css
@vars  { baseFontSize:30 }
body   { fontSize: baseFontSize   }
header { fontSize: baseFontSize*2 }
```

Loading multiple stylesheets into the same style allows later stylesheets to use variables defined in an earlier stylesheet. Further, disabling and enabling a stylesheet disables its variables. Combining these allows you to load and control 'theme' stylesheets. For example:

**day.css**

```css
@vars {
  background: 'white';
  foreground: 'black';
}
```

**night.css**

```css
@vars {
  background: 'darkblue';
  foreground: 'lightblue';
}
```

**content.css**

```css
body { bgColor:background; textColor:foreground }
```

```lua
local body = {type='body'}
local style = zss:new{ files={'day.css', 'night.css', 'content.css'} }
print(style:match(body).bgColor) --> darkblue

style:disable('night.css')
print(style:match(body).bgColor) --> white

style:disable('day.css')
print(style:match(body).bgColor) --> nil
```

Multiple `@vars { … }` directives may appear anywhere in a stylesheet, and safely rely on variables from previously-loaded stylesheets as well as [constants](#myzssconstantsvalue_map) set for the style. However, the following behavior is undefined:

* Placing a rule that uses a variable before the `@vars { … }` directive that defines the variable.

    ```css
    /* this rule should be after the @vars */
    bad { size:x }
    @vars { x:20 }
    ```


* Referencing the value of a variable defined in the same stylesheet:

    ```css
    @vars {
       x : 25;
       y : x + 17; /* regardless of order, do not rely on another variable… */
    }
    @vars {
       z : x * 2; /* …not even in a later @vars block in the same sheet */
    }
    ```

    _A future release may allow later @vars blocks to refer to earlier blocks reliably._



# API

* [`ZSS:new()`](#zssnewopts) — create a style (aggregating many sheets and rules)
* [`myZSS:constants()`](#myzssconstantsvalue_map) — set value lookups
* [`myZSS:directives()`](#myzssdirectiveshandlers) — set at-rule handlers
* [`myZSS:add()`](#myzssaddcss) — parse CSS from string
* [`myZSS:load()`](#myzssload) — load CSS from file
* [`myZSS:match()`](#myzssmatchelement) — compute the declarations that apply to an element
* [`myZSS:extend()`](#myzssextend) — create a new style that derives from this one
* [`myZSS:disable()`](#myzssdisablesheetid) — stop using rules from a particular set of CSS
* [`myZSS:enable()`](#myzssenablesheetid) — resume using rules from a particular set of CSS


## ZSS:new(opts)

`opts` is a table with any of the following string keys (all optional):

* `constants` — a table mapping string literals (that might appear as values or functions in declarations) to the Lua value you would like them to have instead.
* `directives` — a table mapping the string name of an "at-rule" (without the `@`) to a function that should be invoked each time it is seen.
* `basecss` — a string of CSS rules to initially parse for the style.
* `files` — an array of string file names to load and parse (after the `basecss`).

### Example:

```lua
URL  = require'urlhandler'
FONT = require'myfontlib'
ZSS  = require'zss'
local style = ZSS:new{
   constants = {
      none  = false,
      color = require('color'),
      url   = URL.processURL,
   },
   directives = {
      -- e.g. @font-face { font-family='bold'; src:'fonts/Arial-bold.ttf' }
      ['font-face'] = function(style, props)
         FONT.createFont(props['font-family'], props.src)
      end
   },
   basecss = [[...css code...]],
   files   = {'a.css', 'b.css'},
}
```

Constants can be added to later using the `constants()` method.
Additional directives can be added later using the `directives()` method.
Additional raw CSS can be parsed later using the `add()` method.
Additional CSS files can be loaded by name later using the `load()` method.


## myZSS:constants(value_map)

Add key/value pairs to the literal value mappings used when parsing declaration values.

`value_map` is a table mapping string literals (that might appear as values or functions in declarations) to the Lua value you would like them to have instead.

The return value is the invoking ZSS style instance (for method chaining).

**Note**: constants are resolved when rules are added/loaded, unless there is a deferred reference in the value (e.g. `@foo`). Invoking the `constants()` method after CSS has been loaded will only affect CSS added/loaded after the call…or the evaluation of deferred values.

### Example:

```lua
local style = ZSS:new()
style:constants{ none=false, white={1,1,1}, red={1,0,0} }
style:add('* { a:none; b:white; c:red; d:orange } ')
style:match{ id='foo' }
--> {a=false, b={1,1,1}, c={1,0,0}, d=nil}
```

## myZSS:directives(handlers)

Add to the handler functions used when an at-rule is encountered during parsing.

`handlers` is a table mapping the name of the at-rule (without the leading `@`) to a Lua function to invoke. The function will be passed a reference to the ZSS style instance, and a table mapping property names to values seen in the declaration.

The return value is the invoking ZSS style instance (for method chaining).

### Example:

```lua
local myfontlib = require'somefontlib'
local style = ZSS:new()
style:directives{
  font = function(me, props)
    me:constants{ [props.name]=myfontlib.loadFont(props.src) }
  end
}
style:add[[
  @font { name:'mainFont'; src:'my.ttf' }
  text { font:mainFont }
]]
```


## myZSS:add(css)

Add CSS rules to the style (specified as a string of CSS).

**Notes**:

* You should invoke this method only after setting up any `constants` and `directives` mappings for the style.
* Invoking this method resets the computed rules cache (as new rules may invalidate cached computations).

### Example:

```lua
local style = ZSS:new()
style:add[[
  /* CSS comments are recognized and ignored */
  @some-rule { note:'at rules are supported' }
  #special   { x:1; y:2.0; z:'three'; q:sum(1,2,1) }
]]
```

The first return value is the invoking ZSS style instance (for method chaining).

The second return value is a unique identifier that may be used to later `disable()` and `enable()` the rules loaded in this CSS.


## myZSS:load(...)

Add CSS rules to the style, loaded from one or more files specified by file name.

### Example:

```lua
local style = ZSS:new()
style:constants(myconstants)
style:load('a.css', 'b.css', 'c.css')
```


## myZSS:match(element)

Use all rules in the style to compute the declarations that apply to described element.

`element` must be a table using any/all/none of the keys `type`, `id`, and `tags`, and may include arbitrary additional data properties:
`{type='mytype', id='myid', tags={tag1=1, tag2=1}, var1=42, var2=3.8}`.


### Example:

```lua
local style = ZSS:new():add[[
  *          { fill:none;  stroke:white; opacity:1.0 }
  text       { fill:white; stroke:none }
  .important { weight:bold; fill:red }
  .debug     { opacity:0.5 }
]]
local m1 = style:match{ type='text', tags={important=1} }
--> { fill='red', opacity=1.0, stroke='none', weight='bold' }

local m2 = style:match{ type='line', tags={debug=1} }
--> { fill='none', opacity=0.5, stroke='white' }

local m3 = style:match{}
--> { fill='none', opacity=1.0, stroke='white' }
```


## myZSS:extend(...)

Creates a new ZSS style that derives from this style, while retaining a live link to its state. Rules in the new style supercede rules in the 'parent' style (for the same rank). If you pass any filenames, they will be passed to `load()`. (In other words, `local b = a:extend('a.css', 'b.css')` is a shorthand for `local b = a:extend():load('a.css', 'b.css')`.)

### Example:

```lua
local main   = ZSS:new{ files={'base.css', 'day.css'} }
local style1 = main:extend():load('style1.css') -- style1:match() uses base.css, day.css, and style1.css
local style2 = main:extend('style2.css')        -- style2:match() uses base.css, day.css, and style2.css
main:disable('day.css')
main:load('night.css')
--> now style1:match() uses base.css, **night.css**, and style1.css
```

## myZSS:disable(sheetid)

Prevent a specific set of CSS from being used in the style (and all descendants created via `clone()`). For files, the `sheetid` is the path passed to `load()`; for CSS strings added via the `add()` method, the `sheetid` is the second string returned from `add()`. For css added using the `basecss` option to `new()`, use the string `"basecss"`.

### Example:

```lua
local main   = ZSS:new{ files={'base.css', 'day.css'} }
local style1 = main:clone():load('style1a.css')
local _, bID = style1:add('#additional { rules:here }')

style1:match(…) --> uses base.css, day.css, style1a.css, and additional rules loaded during add()

style1:disable(bID)
style1:match(…) --> uses base.css, day.css, style1a.css

main:disable('day.css')
style1:match(…) --> uses base.css and style1a.css
main:match(…)   --> uses base.css only

main:enable('day.css')
style1:match(…) --> uses base.css, day.css, and style1a.css
main:match(…)   --> uses base.css and day.css

style1:disable('day.css')
style1:match(…) --> uses base.css and doc1a.css
main:match(…)   --> uses base.css and day.css
```

**Note**: As shown in the last lines of the example above, a cloned style may disable rules _loaded_ in its parent; however, doing so does not disable them for the parent (or other clones of the parent).

## myZSS:enable(sheetid)

Re-enable CSS previously disabled with `disable()`. (See `disable()` for an example.)


# License & Contact

ZSS is copyright ©2018–2019 by Gavin Kistner and is licensed under the [MIT License](https://opensource.org/licenses/MIT). See the LICENSE.txt file for more details.

For bugs or feature requests please [open issues on GitHub](https://github.com/Phrogz/ZSS/issues). For other communication you can [email the author directly](mailto:!@phrogz.net?subject=ZSS).


HISTORY.md
# History

## v1.0.1 — September 17th, 2019

* Speed improvement to matching (2✕ in the common case)


## v1.0 — August 29th, 2019

### New Features
* Conditions in CSS selectors may now use arbitrary Lua expressions,
  referencing attributes of the element, constants in the sheet, and full operators.
  * Old: `foo[@bar<17], foo[@bar=17] { /* hack around lack of <= support */ }`
  * New: `foo[17>=@bar][@baz<sheetconstant][@jim~=true] { /* so much power */ }`

### Backwards-Incompatible Changes
* All _matching_ takes place using only table 'elements' instead of 'descriptor' strings.
  * Old: `style:match('foo#bar.jim.jam[@v1=17][@v2=42]')`
  * New: `style:match{type='foo', id='bar', tags={jim=1, jam=1}, v1=17, v2=42}`
  * This update removes any computed match caching.
    There isn't a convenient string to hang the cache off of,
    and caching based on element tables will break if the table is mutated.
* All property-equality tests in conditions CSS selectors must now use two equals signs,
  as a single equals sign in a Lua expression is an assignment instead of a test.
  * Old: `foo[@bar=17] { … }`
  * New: `foo[@bar==17] { … }`
* All property names in conditions in CSS selectors must be prefixed with an `@` sign,
  now that the conditions support arbitrary expressions and may reference sheet constants.
  * Old: `foo[bar>42] { … }`
  * New: `foo[@bar<42] { … }`


test.lua
local i = require'inspect'
function p(...) for _,o in ipairs{...} do print(i(o)) end end
p(z)

ZSS = require 'zss'

local function color(a,b,c)
	print('COLOR!',a,b,c)
	return 'COLOR'
end

-- parse raw CSS from a string
rules = ZSS:new{
	values   = { none=false, ['false']=false, ['true']=true },
	handlers = { color=color, rgba=color },
	basecss  = [[
		@font-face { font-family:'main'; src:'DINPro.otf';  }
		@font-face { font-family:'bold'; src:'DINPro-Bold.otf' }

	  /* Make things stand out as ugly, to encourage adding necessary data */
		*    { fill:none; stroke:'magenta';
		       stroke-width:1; font:'main' }

		text { font:'main'; fill:color('white'); stroke:none; size:12; opacity:0.5 }

		text.detection { size:8; fill:green }
		.detection { fill:color('#ffff0033'); stroke:'yellow'; stroke-width:2 }

		.danger { fill:rgba(2,0,0); size:18; effect:glow(yellow,3) }
		.pedestrian { fill:purple; fill-opacity:0.3 }
		.child      { fill-opacity:0.8 }

		*[ttc<1.5]  { effect:flash(0.2) }
		bbox { fill:none; stroke-opacity:0.2 }
		*[speed] { speed:true }
		#accel { stroke:color('green') }
	]]
}

-- add a second document (can also load from a file)
rules:add [[
	@font-face { font-family:'main'; src:'DINPro-Light.otf';  }
	@font-face { font-family:'bold'; src:'DINPro-Medium.otf'; }
	text { fill:orange }
]]

p( rules:match'line.child.counted' )
-- p( rules:match{ id='accl', type='text', tags={danger=1,debug=1}, data={speed=5} } )


test/lerp.lua
local lerp = {}

-- Linear lerp function of the form y=mx + b
-- Usage:
--   local f,b = lerp.linear{in1=1, out1=4, in2=100, out2=1, clamp=true}
--   print("The y-intercept is", b)
--   for x=1,100,10 do
--     print("At", x, "the output is", f(x))
--   end
function lerp.linear(inputs)
	local m = (inputs.out2-inputs.out1)/(inputs.in2-inputs.in1)
	local b = inputs.out1 - m*inputs.in1
	if inputs.clamp then
		local min,max = inputs.in1, inputs.in2
		if min>max then min,max = max,min end
		return function(x)
			x = x<min and min or x>max and max or x
			return m*x + b
		end
	else
		return function(x) return m*x + b end
	end
end

-- Asymptotic lerp function of the form y=s/x + b
-- Usage:
--   local width,min = lerp.asymptote{in1=1, out1=4, in2=100, out2=1, clamp=true}
--   print("The asymptote will bottom out at", min)
--   for x=1,100,10 do
--     print("At", x, "the output is", width(x))
--   end
function lerp.asymptote(inputs)
	local s = inputs.in1*inputs.in2*(inputs.out2-inputs.out1) / (inputs.in1-inputs.in2)
	local b = inputs.out1 - s/inputs.in1
	if inputs.clamp then
		local min,max = inputs.in1, inputs.in2
		if min>max then min,max = max,min end
		return function(x)
			x = x<min and min or x>max and max or x
			return s/x + b
		end, b
	else
		return function(x) return s/x + b end, b
	end
end

return lerp

test/test.lua
local i=function(x) print(require('inspect')(x)) end
package.path = '../?.lua;' .. package.path
zss = require'zss'
ZSS = zss
color = require'color'

local style = ZSS:new()
style:constants{lerp=lerp}:directives{ vars = function(me, props) me:constants(props) end }
style:add[[
  @vars   { uiscale:1.5; fg:'white'; dlerp:lerp.linear{in1=0, in2=10, out1=100, out2=200}; yes:'very' }
  text    { font-size:12*uiscale; fill:fg }
  foo[@x] { bar:dlerp(@x); yes:yes }
]]
i(style:match 'foo[x=3]')


zss.lua
--[=========================================================================[
   ZSS v1.0.1
   See http://github.com/Phrogz/ZSS for usage documentation.
   Licensed under MIT License.
   See https://opensource.org/licenses/MIT for details.
--]=========================================================================]

local ZSS = { VERSION="1.0.1", debug=print, info=print, warn=print, error=print, ANY_VALUE={}, NIL_VALUE={} }

local updaterules, updateconstantschain, dirtyblocks

-- ZSS:new{
--   constants  = { none=false, transparent=processColor(0,0,0,0), color=processColor },
--   directives = { ['font-face']=processFontFaceRule },
--   basecss    = [[...css code...]],
--   files      = { 'a.css', 'b.css' },
-- }
function ZSS:new(opts)
	local style = {
		_directives  = {}, -- map from name of at rule (without @) to function to invoke
		_constants   = {}, -- value literal strings mapped to equivalent values (e.g. "none"=false)
		_sheets      = {}, -- array of sheetids, also mapping sheetid to its active state
		_sheetconst  = {}, -- map of sheetid to table of constants for that sheet
		_sheetblocks = {}, -- map of sheetid to table of blocks underlying the constants
		_envs        = {}, -- map of sheetid to table that mutates to evaluate functions
		_rules       = {}, -- map of sheetids to array of rule tables, each sorted by document order
		_lookup      = {}, -- rules for active sheets indexed by type, then by id (with 'false' indicating no type or id)
		_kids        = setmetatable({},{__mode='k'}), -- set of child tables mapped to true
		_parent      = nil, -- reference to the style instance that spawned this one
	}
	style._envs[1] = setmetatable({},{__index=style._constants})
	setmetatable(style,{__index=self})
	style:directives{
		-- Process @vars { foo:42 } with sheet ordering and chaining
		vars = function(self, values, sheetid, blocks)
			local consts = self._sheetconst[sheetid]
			local blocksforthissheet = self._sheetblocks[sheetid]
			if not blocksforthissheet then
				blocksforthissheet = {}
				self._sheetblocks[sheetid] = blocksforthissheet
			end
			-- iterate over blocks instead of values, in case a value is nil
			for k,v in pairs(blocks) do
				consts[k] = values[k]
				blocksforthissheet[k] = blocks[k]
			end
		end
	}

	if opts then
		if opts.constants  then style:constants(opts.constants,true) end
		if opts.directives then style:directives(opts.directives)    end
		if opts.basecss    then style:add(opts.basecss, 'basecss')   end
		if opts.files      then style:load(table.unpack(opts.files)) end
	end

	updaterules(style)
	return style
end

-- Usage: myZSS:load( 'file1.css', 'file2.css', ... )
function ZSS:load(...)
	self._deferupdate = true
	for _,filename in ipairs{...} do
		local file = io.open(filename)
		if file then
			self:add(file:read('*all'), filename)
			file:close()
		else
			error("Could not load CSS file '"..filename.."'.")
		end
	end
	updaterules(self)
	self._deferupdate = false
	return self
end

-- Usage: myZSS:constants{ none=false, transparent=color(0,0,0,0), rgba=color.makeRGBA }
function ZSS:constants(valuemap, preserveCaches)
	for k,v in pairs(valuemap) do self._constants[k]=v end
	if not preserveCaches then dirtyblocks(self) end
	return self
end

function ZSS:directives(handlermap)
	for k,v in pairs(handlermap) do self._directives[k]=v end
	return self
end

-- Turn Lua code into function blocks that return the value
function ZSS:compile(str, sheetid, property)
	local dynamic = str:find('@[%a_][%w_]*') or str:find('^!')
	local env = self._envs[sheetid] or self._envs[1]
	local func,err = load('return '..str:gsub('@([%a_][%w_]*)','_data_.%1'):gsub('^!',''), str, 't', env)
	if not func then
		self.error(("Error compiling CSS expression for '%s' in '%s': %s"):format(property, sheetid, err))
	else
		-- [1] will hold the actual cached value (deferred until the first request)
		-- [2] is whether or not a cached value exists in [1] (needed in case the computed value is nil)
		-- [3] is the environment of the function (cached for convenient lookup)
		-- [4] is the function that produces the value
		-- [5] indicates if the block must be computed each time, or may be cached
		-- [6] is the sheetid (used for evaluation error messages)
		-- [7] is the name of the property being evaluated (used for evaluation error messages)
		return {nil,false,env,func,dynamic,sheetid,property,zssblock=true}
	end
end

-- Compute the value for a block
function ZSS:eval(block, data, ignorecache)
	if block[2] and not ignorecache then return block[1] end
	block[3]._data_ = data -- may be nil; important to clear out from previous eval
	local ok,valOrError = pcall(block[4])
	if ok then
		if not block[5] then
			block[2]=true
			block[1]=valOrError
		end
		return valOrError
	else
		self.error(("Error evaluating CSS value for '%s' in '%s': %s"):format(block[7], block[6], valOrError))
	end
end

-- Convert a selector string into its component pieces;
function ZSS:parse_selector(selector_str, sheetid)
	local selector = {
		type = selector_str:match('^@?[%a_][%w_-]*') or false,
		id  = selector_str:match('#([%a_][%w_-]*)') or false,
		tags={}, data={}
	}
	local tagrank = 0

	-- Find all the tags
	for name in selector_str:gmatch('%.([%a_][%w_-]*)') do
		selector.tags[name] = true
		tagrank = tagrank + 1
	end

	-- Find all attribute sections, e.g. [@attr], [@attr<17], [attr=12], …
	for attr in selector_str:gmatch('%[%s*(.-)%s*%]') do
		local attr_name_only = attr:match('^@?([%a_][%w_-]*)$')
		if attr_name_only then
			selector.data[attr_name_only] = ZSS.ANY_VALUE
			tagrank = tagrank + 1
		else
			local name, op, val = attr:match('^@([%a_][%w_-]*)%s*(==)%s*(.-)$')
			if name then
				local value = self:eval(self:compile(val, sheetid, selector_str))
				selector.data[name] = value==nil and ZSS.NIL_VALUE or value
			else
				selector.data[attr] = self:compile(attr, sheetid, selector_str)
			end
			-- attribute selectors with comparisons count slightly more than bare attributes or tags
			tagrank = tagrank + 1.001
		end
	end

	selector.rank = {
		selector.id and 1 or 0,
		tagrank,
		selector.type and 1 or 0,
		0 -- the document order will be determined during updaterules()
	}

	return selector
end

-- Add a block of raw CSS rules (as a single string) to the style sheet
-- Returns the sheet itself (for chaining) and id associated with the css (for later enable/disable)
function ZSS:add(css, sheetid)
	sheetid = sheetid or 'css#'..(#self:sheetids()+1)

	local newsheet = not self._rules[sheetid]
	if newsheet then
		table.insert(self._sheets, sheetid)
		self._sheets[sheetid] = true
		self._sheetconst[sheetid] = setmetatable({},{__index=self._constants})
		self._envs[sheetid] = setmetatable({},{__index=self._sheetconst[sheetid]})
		-- inherit from the last active sheet in the chain
		for i=#self._sheets-1,1,-1 do
			if self._sheets[self._sheets[i]] then
				getmetatable(self._sheetconst[sheetid]).__index = self._sheetconst[self._sheets[i]]
				break
			end
		end
	end
	self._rules[sheetid] = {}

	for rule_str in css:gsub('/%*.-%*/',''):gmatch('[^%s][^{]+%b{}') do
		-- Convert declarations into a table mapping property to block
		local decl_str = rule_str:match('[^{]*(%b{})'):sub(2,-2)
		local blocks = {}
		for key,val in decl_str:gmatch('([^%s:;]+)%s*:%s*([^;]+)') do
			blocks[key] = self:compile(val, sheetid, key)
		end

		-- Create a unique rule for each selector in the rule
		local selectors_str = rule_str:match('(.-)%s*{')
		for selector_str in selectors_str:gmatch('%s*([^,]+)') do
			-- Check if this is a directive (at-rule) that needs processing
			local name = selector_str:match('^%s*@([%a_][%w_-]*)%s*$')
			if name and self._directives[name] then
				-- bake value blocks into values before handing off
				local values = {}
				for k,block in pairs(blocks) do values[k] = self:eval(block) end
				self._directives[name](self, values, sheetid, blocks)
			end

			local selector = self:parse_selector(selector_str:match "^%s*(.-)%s*$", sheetid)
			if selector then
				local rule = {selector=selector, declarations=blocks, doc=sheetid, selectorstr=selector_str}
				table.insert(self._rules[sheetid], rule)
			end
		end
	end

	-- sort rules and determine active based on rank
	if not self._deferupdate then updaterules(self) end

	return self, sheetid
end

-- Given an element table, compute the declaration(s) that apply. For example:
-- local values = myZSS:match{tags={ped=1}}
-- local values = myZSS:match{type='text', id='accel', tags={label=1}}
-- local values = myZSS:match{type='text', id='accel', tags={value=1, negative=1}, value=-0.3}
function ZSS:match(el)
	local placeholders, data

	local result, sortedrules, ct = {}, {}, 0

	local function checkrules(rules)
		for _,rule in ipairs(rules) do
			local sel = rule.selector
			for tag,_ in pairs(sel.tags) do
				-- TODO: handle anti-tags in the selector, where _==false
				if not (el.tags and el.tags[tag]) then goto skiprule end
			end
			for name,desired in pairs(sel.data) do
				if type(desired)=='table' and desired.zssblock then
					if not self:eval(desired, el, true) then goto skiprule end
				else
					local actual = el[name]
					if desired==ZSS.NIL_VALUE then
						if actual~=nil then goto skiprule end
					elseif desired==ZSS.ANY_VALUE then
						if actual==nil then goto skiprule end
					else
						if actual~=desired then goto skiprule end
					end
				end
			end
			ct = ct + 1
			sortedrules[ct] = rule
			::skiprule::
		end
	end

	local function addfortype(byid)
		if el.id and byid[el.id] then addforid(byid[el.id]) end
		if byid[false] then addforid(byid[false]) end
	end

	-- check rules that specify this element's type
	local byid = el.type and self._lookup[el.type]
	if byid then
		-- check rules that specify this element's id
		if el.id and byid[el.id] then checkrules(byid[el.id]) end

		-- check rules that don't care about the element id
		if byid[false] then checkrules(byid[false]) end
	end

	-- check rules that don't care about the element type
	local byid = self._lookup[false]
	if byid then
		-- check rules that specify this element's id
		if el.id and byid[el.id] then checkrules(byid[el.id]) end

		-- check rules that don't care about the element id
		if byid[false] then checkrules(byid[false]) end
	end

	if ct>1 then
		table.sort(sortedrules, function(r1, r2)
			r1,r2 = r1.selector.rank,r2.selector.rank
			if     r1[1]<r2[1] then return true elseif r1[1]>r2[1] then return false
			elseif r1[2]<r2[2] then return true elseif r1[2]>r2[2] then return false
			elseif r1[3]<r2[3] then return true elseif r1[3]>r2[3] then return false
			elseif r1[4]<r2[4] then return true else                    return false
			end
		end)
	end

	-- merge the declarations from the rules in order
	for _,rule in ipairs(sortedrules) do
		for k,block in pairs(rule.declarations) do
			result[k] = block
		end
	end

	-- Convert blocks to values (honoring cached values)
	for k,block in pairs(result) do
		result[k] = self:eval(block, el)
	end

	return result
end

function ZSS:extend(...)
	local kid = self:new()
	kid._parent = self
	-- getmetatable(kid).__index = self
	setmetatable(kid._constants, {__index=self._constants})
	setmetatable(kid._directives,{__index=self._directives})
	setmetatable(kid._envs,{__index=self._envs})
	self._kids[kid] = true
	kid:load(...)
	return kid
end

function ZSS:sheetids()
	local ids = self._parent and self._parent:sheetids() or {}
	table.move(self._sheets, 1, #self._sheets, #ids+1, ids)
	for _,id in ipairs(self._sheets) do ids[id] = true end
	return ids
end

function ZSS:disable(sheetid)
	local ids = self:sheetids()
	if ids[sheetid] then
		self._sheets[sheetid] = false
		updaterules(self)
		updateconstantschain(self)
		dirtyblocks(self,sheetid)
	else
		local quote='%q'
		for i,id in ipairs(ids) do ids[i] = quote:format(id) end
		self.warn(("Cannot disable() CSS with id '%s' (no such sheet loaded).\nAvailable sheet ids: %s"):format(sheetid, table.concat(ids, ", ")))
	end
end

function ZSS:enable(sheetid)
	if self._rules[sheetid] then
		self._sheets[sheetid] = true
		updaterules(self)
		updateconstantschain(self)
		dirtyblocks(self,sheetid)
	elseif self._sheets[sheetid]~=nil then
		-- wipe out a local override that may have been disabling an ancestor
		self._sheets[sheetid] = nil
		updaterules(self)
		updateconstantschain(self)
		dirtyblocks(self,sheetid)
	else
		local disabled = {}
		local quote='%q'
		for id,state in pairs(self._sheets) do
			if state==false then table.insert(disabled, quote:format(id)) end
		end
		if #disabled==0 then
			self.warn(("Cannot enable() CSS with id '%s' (no sheets are disabled)."):format(sheetid))
		else
			self.warn(("Cannot enable() CSS with id '%s' (no such sheet disabled).\nDisabled ids: %s"):format(sheetid, table.concat(disabled, ", ")))
		end
	end
end

updaterules = function(self)
	-- reset the lookup cache
	self._lookup = {}
	local rulesadded = 0

	-- add all rules from the parent style
	if self._parent then
		for type,rulesfortype in pairs(self._parent._lookup) do
			local byid = self._lookup[type]
			if not byid then
				byid = {}
				self._lookup[type] = byid
			end
			for id,rulesforid in pairs(rulesfortype) do
				local rules = byid[id]
				if not rules then
					rules = {}
					byid[id] = rules
				end
				for i=1,#rulesforid do
					local rule = rulesforid[i]
					-- do not use rule if we set its document inactive
					if self._sheets[rule.doc]~=false then
						rules[#rules+1] = rule
						rulesadded = rulesadded + 1
					end
				end
			end
		end
	end

	-- add all the rules from this style's active sheets
	for _,sheetid in ipairs(self._sheets) do
		if self._sheets[sheetid] then
			for _,rule in ipairs(self._rules[sheetid]) do
				local byid = self._lookup[rule.selector.type]
				if not byid then
					byid = {}
					self._lookup[rule.selector.type] = byid
				end
				local rulesforid = byid[rule.selector.id]
				if not rulesforid then
					rulesforid = {}
					byid[rule.selector.id] = rulesforid
				end
				rulesforid[#rulesforid+1] = rule
				rulesadded = rulesadded + 1
				rule.selector.rank[4] = rulesadded
			end
		end
	end

	-- ensure that any extended children are updated
	for kid,_ in pairs(self._kids) do
		updaterules(kid)
	end
end

updateconstantschain = function(self)
	local lastactive = self._constants
	for _,sheetid in ipairs(self._sheets) do
		-- If the sheet is active, put it into the chain
		if self._sheets[sheetid] then
			local sheetconst = self._sheetconst[sheetid]
			getmetatable(sheetconst).__index = lastactive
			lastactive = sheetconst
		end
	end
end

dirtyblocks = function(self, sheetid)
	local dirtythissheet = not sheetid
	for _,id in ipairs(self._sheets) do
		if id==sheetid then dirtythissheet=true end
		if dirtythissheet then
			for _,rule in ipairs(self._rules[id]) do
				for k,block in pairs(rule.declarations) do block[2]=false end
			end
			if self._sheetblocks[id] then
				for k,block in pairs(self._sheetblocks[id]) do
					self._sheetconst[id][k] = self:eval(block,nil,true)
				end
			end
		end
	end
	for kid,_ in pairs(self._kids) do dirtyblocks(kid) end
end

return ZSS
